src/transformStateWithClones.js

```diff
@@ -7,7 +7,25 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const newArray = [];
+
+  let object = { ...state };
+
+  for (const action of actions) {
+    if (action.type === 'addProperties') {
+      object = { ...object, ...action.extraData };
+    } else if (action.type === 'removeProperties') {
+      for (const key of action.keysToRemove) {
+        delete object[key];
+      }
+    } else if (action.type === 'clear') {
+      object = {};
+    }
+
+    newArray.push({ ...object });
```

Comment:
Pushing a copy of the object to the array at the end of each loop cycle is correct, but ensure that it is done outside of the 'switch' block, after the current action is processed.